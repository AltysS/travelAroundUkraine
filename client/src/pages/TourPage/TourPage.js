import {
  alpha,
  Box,
  Button,
  Container,
  Dialog,
  Link,
  Slide,
  Stack,
  styled,
  Typography,
  useMediaQuery,
} from '@mui/material';
import React, { useState } from 'react';
// Icons
import BedIcon from '@mui/icons-material/Bed';
import PersonIcon from '@mui/icons-material/Person';

import { useInView } from 'react-intersection-observer';

import { TourAccordion, TourInfoDialog, TourReasonToChoose } from '../../components';

const sections = [
  { title: 'About us', link: '#about-tour' },
  { title: 'Reasons to choose', link: '#reasons-to-choose' },
  { title: 'What is included?', link: '#included' },
];

const currency = {
  eur: '€',
  usd: '$',
  uah: '₴',
};

const included = [
  { icon: <PersonIcon color="primary" />, service: 'Professional guide' },
  { icon: <BedIcon color="primary" />, service: 'Accomodation' },
];

const cost = {
  eur: 70,
  usd: 70,
  uah: 2584,
};

const dates = { beginDate: new Date('2022-02-01'), endDate: new Date('2022-02-25') };

const details = {
  duration: '3 days',
  departs: 'First city',
  returns: 'Second city',
};

const HeaderContent = styled(Box)(({ theme }) => ({
  [theme.breakpoints.up('laptop')]: {
    borderBottom: `1px solid ${theme.palette.divider}`,
  },
}));

const Nav = styled('nav')(({ theme }) => ({
  paddingBlock: '15px',

  [theme.breakpoints.up('laptop')]: {
    paddingBlock: '20px',
  },
}));

// eslint-disable-next-line react/jsx-props-no-spreading
const NavLink = styled((props) => <Link variant="body1" underline="none" {...props} />)(({ theme }) => ({
  color: alpha(theme.palette.text.primary, 0.5),

  transition: theme.transitions.create('color'),

  '&:hover': {
    color: theme.palette.text.primary,
  },

  [theme.breakpoints.up('tablet')]: {
    fontSize: '16px',
  },
}));

const LinksWrapper = styled(Stack)(({ theme }) => ({
  flexDirection: 'row',
  gap: '20px',

  [theme.breakpoints.up('tablet')]: {
    justifyContent: 'center',
  },
}));

const MainContent = styled(Box)(({ theme }) => ({
  paddingBlock: '10px',

  [theme.breakpoints.up('laptop')]: {
    paddingBlock: '25px',
  },
}));

const ContentWrapper = styled(Stack)({
  flexDirection: 'row',
  gap: '20px',
});

const Section = styled(Box)({
  scrollMarginTop: '20px',
});

const MobileDialogWrapper = styled(Box)(({ theme }) => ({
  paddingBlock: '10px',
  borderBlock: `1px solid ${theme.palette.divider}`,
}));

const Cost = styled(Typography)(({ theme }) => ({
  fontWeight: 700,
  marginBottom: 0,
  fontSize: 25,
  lineHeight: 1,
  color: theme.palette.primary.main,
}));

const FloatingDialog = styled(Box)(({ theme }) => ({
  position: 'fixed',
  width: '100%',
  left: 0,
  bottom: 0,
  borderTop: `1px solid ${theme.palette.divider}`,
  backgroundColor: '#fff',
  paddingBlock: '15px',
  zIndex: '1',
}));

const TourPage = () => {
  const matchesMediaQuery = useMediaQuery('(min-width: 1051px)');

  const [isOpen, setIsOpen] = useState(false);

  const [dialogRef, inView] = useInView({ root: null, rootMargin: '0px', threshold: 0.1 });

  const handleOpenDialog = () => setIsOpen(true);
  const handleCloseDialog = () => setIsOpen(false);

  return (
    <>
      <HeaderContent>
        <Container>
          <Nav>
            <LinksWrapper>
              {sections.map(({ title, link }) => (
                <NavLink key={title} href={link}>
                  {title}
                </NavLink>
              ))}
            </LinksWrapper>
          </Nav>
        </Container>
      </HeaderContent>

      <MainContent>
        <Container>
          <ContentWrapper>
            {matchesMediaQuery ? (
              <Box component="aside" sx={{ maxWidth: '370px', width: '100%' }}>
                <TourInfoDialog included={included} cost={cost} dates={dates} details={details} />
              </Box>
            ) : null}

            <Box sx={{ flex: 1 }}>
              <Section id="about-tour">
                <Typography variant="h2">About tour</Typography>
                <Box sx={{ paddingBottom: 3 }}>
                  <Typography>
                    Each medieval city is full of various legends and mystical stories. Lviv is no exception because
                    more than 760 years of history simply could not help but leave mysterious legends.
                  </Typography>
                  <Typography>
                    The guide will tell and show that Lviv legends can impress no less than Lviv architecture. This tour
                    will be filled with incredible stories - unusual stories from the life of the Galician capital.
                  </Typography>
                  <Typography>
                    How can one explain the black colour of one of the houses on Rynok Square? How did Adam Senyavsky
                    marry his daughter? How did the first bulletin board appear in Lviv? Why did one Lviv lady make a
                    not entirely decent offer to the tram? Are the legends about Lviv ghosts true? What do Lviv
                    courtyards hide?
                  </Typography>
                  <Typography>All this and more you can learn during this tour.</Typography>
                  <Typography>Looks like today. The program of individual excursions: at any hour.</Typography>
                </Box>
              </Section>

              <Section id="reasons-to-choose">
                <TourAccordion id="reasons-to-choose" title="Reasons to choose our tour">
                  <Stack direction="row" gap={3} flexWrap="wrap" mt={2} pl={2}>
                    <TourReasonToChoose
                      number={1}
                      description="Lorem ipsum dolor sit amet consectetur, adipisicing elit. Nisi ducimus tenetur in aspernatur, asperiores."
                    />
                    <TourReasonToChoose
                      number={2}
                      description="Lorem ipsum dolor sit amet consectetur, adipisicing elit."
                    />
                    <TourReasonToChoose
                      number={3}
                      description="Lorem ipsum dolor sit amet consectetur, adipisicing elit. Nisi ducimus tenetur in aspernatur, asperiores."
                    />
                  </Stack>
                </TourAccordion>
              </Section>

              <Section id="included">
                <TourAccordion title="What is included?">
                  <Stack direction="row" gap="20px">
                    {included.map(({ icon, service }) => (
                      <Stack key={service} direction="row" gap="5px" alignItems="center" flexWrap="wrap">
                        {icon}
                        <Typography gutterBottom={false}>{service}</Typography>
                      </Stack>
                    ))}
                  </Stack>
                </TourAccordion>
              </Section>
            </Box>
          </ContentWrapper>

          {!matchesMediaQuery ? (
            <MobileDialogWrapper ref={dialogRef}>
              <Stack direction="row" justifyContent="space-between" alignItems="center">
                <Cost>
                  {currency.eur}
                  {cost.eur}
                </Cost>
                <Button sx={{ paddingInline: '30px' }} disableElevation onClick={handleOpenDialog}>
                  More info
                </Button>
              </Stack>
              <Dialog open={isOpen} onClose={handleCloseDialog} hideBackdrop fullScreen>
                <TourInfoDialog
                  included={included}
                  cost={cost}
                  dates={dates}
                  details={details}
                  closeButton
                  handleClose={handleCloseDialog}
                />
              </Dialog>
            </MobileDialogWrapper>
          ) : null}
        </Container>

        {!matchesMediaQuery ? (
          <Slide in={!inView} direction="up" mountOnEnter unmountOnExit>
            <FloatingDialog>
              <Container>
                <Stack direction="row" justifyContent="space-between" alignItems="center">
                  <Cost>
                    {currency.eur}
                    {cost.eur}
                  </Cost>
                  <Button sx={{ paddingInline: '30px' }} disableElevation onClick={handleOpenDialog}>
                    More info
                  </Button>
                </Stack>
              </Container>
            </FloatingDialog>
          </Slide>
        ) : null}
      </MainContent>
    </>
  );
};

export default TourPage;
