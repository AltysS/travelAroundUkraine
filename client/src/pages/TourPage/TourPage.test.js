import { render } from '@testing-library/react';
import TourPage from './TourPage';

import { mockAllIsIntersecting } from 'react-intersection-observer/test-utils';

jest.mock('@mui/material/Box', () => 'div');
jest.mock('@mui/material/Stack', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Container', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Typography', () => ({ children }) => <p>{children}</p>);
jest.mock('@mui/material/Button', () => ({ children }) => <button>{children}</button>);
jest.mock('@mui/material/Slide', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Link', () => ({ children }) => <a>{children}</a>);
jest.mock('../../components/TourAccordion/TourAccordion.js', () => ({ children }) => <div>{children}</div>);
jest.mock('../../components/TourInfoDialog/TourInfoDialog.js', () => () => <div>TourInfoDialog</div>);
jest.mock('../../components/TourReasonToChoose/TourReasonToChoose.js', () => () => <div>TourReasonToChoose</div>);

describe('Tour Page snapshot testing', () => {
  test('should Tour Page match snapshot', () => {
    const { asFragment } = render(<TourPage />);
    mockAllIsIntersecting(true);

    expect(asFragment()).toMatchSnapshot();
  });
});
