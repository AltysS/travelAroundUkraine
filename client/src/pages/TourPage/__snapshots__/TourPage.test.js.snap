// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tour Page snapshot testing should Tour Page match snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-2k8nkm"
  >
    <div>
      <nav
        class="css-rtewkn"
      >
        <div>
          <a>
            About us
          </a>
          <a>
            Reasons to choose
          </a>
          <a>
            What is included?
          </a>
        </div>
      </nav>
    </div>
  </div>
  <div
    class="css-souzgl"
  >
    <div>
      <div>
        <div
          sx="[object Object]"
        >
          <div
            class="css-hyfiop"
            id="about-tour"
          >
            <p>
              About tour
            </p>
            <div
              sx="[object Object]"
            >
              <p>
                Each medieval city is full of various legends and mystical stories. Lviv is no exception because more than 760 years of history simply could not help but leave mysterious legends.
              </p>
              <p>
                The guide will tell and show that Lviv legends can impress no less than Lviv architecture. This tour will be filled with incredible stories - unusual stories from the life of the Galician capital.
              </p>
              <p>
                How can one explain the black colour of one of the houses on Rynok Square? How did Adam Senyavsky marry his daughter? How did the first bulletin board appear in Lviv? Why did one Lviv lady make a not entirely decent offer to the tram? Are the legends about Lviv ghosts true? What do Lviv courtyards hide?
              </p>
              <p>
                All this and more you can learn during this tour.
              </p>
              <p>
                Looks like today. The program of individual excursions: at any hour.
              </p>
            </div>
          </div>
          <div
            class="css-hyfiop"
            id="reasons-to-choose"
          >
            <div>
              <div>
                <div>
                  TourReasonToChoose
                </div>
                <div>
                  TourReasonToChoose
                </div>
                <div>
                  TourReasonToChoose
                </div>
              </div>
            </div>
          </div>
          <div
            class="css-hyfiop"
            id="included"
          >
            <div>
              <div>
                <div>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
                    data-testid="PersonIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                    />
                  </svg>
                  <p>
                    Professional guide
                  </p>
                </div>
                <div>
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
                    data-testid="BedIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M21 10.78V8c0-1.65-1.35-3-3-3h-4c-.77 0-1.47.3-2 .78-.53-.48-1.23-.78-2-.78H6C4.35 5 3 6.35 3 8v2.78c-.61.55-1 1.34-1 2.22v6h2v-2h16v2h2v-6c0-.88-.39-1.67-1-2.22zM14 7h4c.55 0 1 .45 1 1v2h-6V8c0-.55.45-1 1-1zM5 8c0-.55.45-1 1-1h4c.55 0 1 .45 1 1v2H5V8z"
                    />
                  </svg>
                  <p>
                    Accomodation
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-13fe9vk"
      >
        <div>
          <p>
            €70
          </p>
          <button>
            More info
          </button>
        </div>
      </div>
    </div>
    <div>
      <div
        class="css-v912gw"
      >
        <div>
          <div>
            <p>
              €70
            </p>
            <button>
              More info
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
